rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,251 @@
+
+# Detailed design
+[design]: #detailed-design
+
+We can break this down nicely into steps.
+
+## Dynamic derivations
+
+*This is implemented in https://github.com/NixOS/nix/pull/4628.*
+
+1. Derivation outputs can be valid derivations:
+
+   1. Allow derivation outputs to be content addressed with the "text hashing" scheme.
+
+   2. Lift the restriction barring derivations output paths from ending in `.drv` if they are so content-addressed
+
+2. Extend the CLI to take advantage of such derivations:
+
+   We hopefully will soon allow CLI "installable" args in the form
+   ```
+   single-installable ::= <path> ! <output-name>
+   ```
+   where the first path is a derivation, and the second is the output we want to build.
+
+   We should generalize the grammar like so:
+   ```
+   single-installable ::= <single-installable> ! <output-name>
+                       |  <path>
+
+   multi-installable  ::= <single-installable>
+                       |  <single-installable> ! *
+   ```
+
+   Plain paths just mean that path itself is the goal, while `!` indexing indicates one more outputs of the derivation to the left of the `!` is the goal.
+
+   > For example,
+   > ```
+   > nix build /nix/store/…foo.drv
+   > ```
+   > would just obtain `/nix/store/…foo.drv` and not build it, while
+   > ```
+   > nix build /nix/store/…foo.drv!*
+   > ```
+   > would obtain (by building or substituting) all its outputs.
+   > ```
+   > nix build /nix/store/…foo.drv!out!out
+   > ```
+   > would obtain the `out` output of whatever derivation `/nix/store/…foo.drv!out` produces.
+
+   Now that we have `path` vs `path!*`, we also don't need `--derivation` as a disambiguator, and so that should be removed along with all the complexity that goes with it.
+   (`toDerivedPathsWithHints` in the the nix commands should always be pure functions and not consult the store.)
+
+3. Extend the scheduler and derivation dependencies similarly:
+
+  - Derivations can depended on the outputs of derivations that are themselves derivation outputs.
+    The schedule will substitute derivations to simplify dependencies as computed derivations are built, just like how floating content addressed derivations are realized.
+
+  - Missing derivations get their own full fledged goals so they can be built, not just fetched from substituters.
+
+4. Add a new `outputOf` primop:
+
+   `builtins.outputOf drv outputName` produces a placeholder string with the appropriate string context to access the output of that name produced by that derivation.
+   The placeholder string is quite analogous to that used for floating content addressed derivation outputs.
+   This isn't needed today when we statically know the derivation and thus it's outputs to make attributes for them, but it is needed for built derivations because we don't know either the final derivation or its output names in advanced.
```

Comment:
I got confused by "isn't needed today", because I wasn't sure if you were saying "we don't need this feature when this RFC is initially accepted", or "we don't need this feature without this RFC". I think you're saying the latter - is that right?

Reply:
I think the whole paragraph can be dropped. Instead:

> It [the placeholder string] holds the not-yet-reified output path for downstream evaluations to continue working with &mdash; without building. In our current model without rhose placeholder strings, the build during eval time is unavoidable.

Reply:
"eval time" is unrelated --- this entire first part of the detail design is store layer only, but I can still reword this to be a lot better.
```suggestion
   \[With just floating content-addressed derivations but no built derivations, derivations are always known statically but their outputs aren't. With this RFC, since derivations themselves can floating CA derivation outputs, we also might not know them statically, so we need "deep" placeholders to account for arbitrary layers of dynamism. This also corresponds to the use of arbitrary many `!` in the CLI.
```

Reply:
Richer! :+1: and a little clearer, too :smile: